src/plugins/intel_cpu/src/emitters/jit_snippets_emitters.cpp

```diff
@@ -0,0 +1,575 @@
+                                     const std::vector<size_t>& out,
+                                     const std::vector<size_t>& vec_pool,
+                                     const std::vector<size_t>& gpr_pool,
+                                     const ov::intel_cpu::emitter_context *emit_context) const {
+    const size_t num_inputs = in[0];
+    const size_t num_outputs = in[1];
+    const size_t vector_size = in[2];
+    const size_t num_params = num_inputs + num_outputs;
+    const size_t outer_work_amount = jcp.scheduler_dims[0];
+    const size_t inner_work_amount = jcp.scheduler_dims[1];
+    const int64_t harness_num_dims = jcp.output_dims.size() - 1;
+
+    const auto& vector_tile = body[0];
+    const auto& scalar_tile = body[1];
+    const auto& vector_tile_body = std::dynamic_pointer_cast<TileEmitter>(vector_tile.first)->get_nested_code();
+    const auto& scalar_tile_body = std::dynamic_pointer_cast<TileEmitter>(scalar_tile.first)->get_nested_code();
+
+    // remove data ptr regs from the pool, since they should be preserved
+    const auto& data_ptr_reg_idxs(out);
+
+    // It is critical that reg_outer_amount and reg_inner_amount represent the
+    // first two runtime arguments, since they are used to calculating offsets
+    Reg64 reg_outer_amount = Reg64(static_cast<int>(in[3]));
+    Reg64 reg_inner_amount = Reg64(static_cast<int>(in[4]));
+
+    Label for_body;
+
+    Reg64 reg_indexes = reg_outer_amount;
+    Reg64 reg_const_params = reg_inner_amount;
+    auto init_ptrs_with_offsets = [&](Reg64 pointer, const int64_t *offsets, Reg64 reg_tmp) {
+        for (int j = 0; j < harness_num_dims; j++) {
+            if (jcp.output_dims[j] != 1 && offsets[j] != 0) {
+                h->mov(reg_tmp, offsets[j]);
+                h->imul(reg_tmp, h->ptr[reg_indexes + j * sizeof(size_t)]);
+                h->add(pointer, reg_tmp);
+            }
+        }
+    };
+    std::vector<Reg64> data_ptr_regs(num_params);
+    for (auto i = 0; i < num_params; i++) {
+        data_ptr_regs[i] = Reg64(static_cast<int>(data_ptr_reg_idxs[i]));
+        if (i < num_inputs)
+            h->mov(data_ptr_regs[i], h->ptr[reg_const_params + GET_OFF(src_ptrs) + i * sizeof(void*)]);
+        else
+            h->mov(data_ptr_regs[i], h->ptr[reg_const_params + GET_OFF(dst_ptrs) + (i - num_inputs) * sizeof(void*)]);
+        // we can use the last data_ptr_reg as tmp_reg until the last iteration, and reg_const_params then
+        Reg64 reg_tmp = i < num_params-1 ? Reg64(static_cast<int>(data_ptr_reg_idxs.back())) : reg_const_params;
+        init_ptrs_with_offsets(data_ptr_regs[i], &jcp.data_offsets[i * harness_num_dims], reg_tmp);
+    }
+    auto emit_tiles = [&]() {
+        bool inner_work_amount_is_set = false;
+        auto process_tile =
+                [&](bool body_condition, const std::vector<EmitterCode>& body,
+                                    bool tile_condition, const EmitterCode& tile) {
```

Comment:
```suggestion
                [&](const bool body_condition, const std::vector<EmitterCode>& body,
                                    const bool tile_condition, const EmitterCode& tile) {
```


Or do we really need two boolean variables? I see they are interchangeable

Reply:
I changed this code significantly, please have a look at the new version